lib/normalize_grammar.re

```diff
@@ -7,56 +7,77 @@ let gensym = () => {
   "intermediate" ++ string_of_int(counter^);
 }
 
-let rec normalize_to_atom = body => {
+let rec normalize_to_simple = (body: A.rule_body): (B.simple, list(A.rule)) => {
   switch (body) {
-  | A.TOKEN => (B.TOKEN, [])
-  | A.SYMBOL(name)=> (B.SYMBOL(name), [])
+  | A.TOKEN | A.IMMEDIATE_TOKEN | A.BLANK | A.SYMBOL(_) | A.STRING(_) | A.PATTERN(_) => {
+    let (atom, rest) = normalize_to_atom(body);
+    (B.ATOM(atom), rest)
+  }
+  | A.SEQ(bodies) => {
+    let xs = List.map(normalize_to_atom, bodies);
+    let atoms = List.map(fst, xs);
+    let intermediates = List.flatten(List.map(snd, xs));
+    (B.SEQ(atoms), intermediates)
+  }
   /* Create intermediate symbol */
   | _ => {
     let fresh_ident = gensym();
-    (B.SYMBOL(fresh_ident), [(fresh_ident, fst(normalize_body(body)))]);
+    (B.ATOM(B.SYMBOL(fresh_ident)), [(fresh_ident, body)]);
     }
   }
 }
-and normalize_body = (rule_body) => {
+and normalize_to_atom = (body: A.rule_body): (B.atom, list(A.rule)) => {
+  switch (body) {
+  | A.TOKEN | A.IMMEDIATE_TOKEN | A.BLANK => (B.TOKEN, [])
+  | A.SYMBOL(name) | A.STRING(name) | A.PATTERN(name)=> (B.SYMBOL(name), [])
+  /* Create intermediate symbol */
+  | _ => {
+    let fresh_ident = gensym();
+    (B.SYMBOL(fresh_ident), [(fresh_ident, body)]);
+    }
+  }
+}
+and normalize_body = (rule_body: A.rule_body): (B.rule_body, list(A.rule)) => {
   switch(rule_body) {
-  | A.TOKEN =>
-    (
-      B.SIMPLE(B.ATOM(B.TOKEN)),
-      []
-    );
-  | A.SYMBOL(name) =>
-    (
-      B.SIMPLE(B.ATOM(B.SYMBOL(name))),
-      []
-    );
+  | A.IMMEDIATE_TOKEN | A.BLANK | A.TOKEN | A.SYMBOL(_) | A.STRING(_) | A.PATTERN(_) => {
+      let (simple, rest) = normalize_to_simple(rule_body);
+      (B.SIMPLE(simple), rest);
+    }
   | A.SEQ(bodies) => {
```

Comment:
You can put SEQ with the other A | B | C above with normalize_to_simple. No need
to duplicate code here.

Reply:
ok